langgraph/src/checkpoint/base.ts

```diff
@@ -65,14 +83,65 @@ export const enum CheckpointAt {
   END_OF_RUN = "end_of_run",
 }
 
-export abstract class BaseCheckpointSaver {
-  at: CheckpointAt = CheckpointAt.END_OF_RUN;
+export interface CheckpointTuple {
+  config: RunnableConfig;
+  checkpoint: Checkpoint;
+  parentConfig?: RunnableConfig;
+}
+
+const CheckpointThreadId: ConfigurableFieldSpec = {
+  id: "threadId",
+  annotation: typeof "",
+  name: "Thread ID",
+  description: null,
+  default: "",
+  isShared: true,
+  dependencies: null,
+};
+
+const CheckpointThreadTs: ConfigurableFieldSpec = {
+  id: "threadTs",
+  annotation: typeof "",
+  name: "Thread Timestamp",
+  description:
+    "Pass to fetch a past checkpoint. If None, fetches the latest checkpoint.",
+  default: null,
+  isShared: true,
+  dependencies: null,
+};
+
+export interface SerializerProtocol<D, L> {
+  dumps(obj: D): L;
+  loads(data: L): D;
+}
+
+export abstract class BaseCheckpointSaver<D, L> {
```

Comment:
D here should always be Checkpoint

Reply:
Makes sense. I removed `D` and hardcoded `Checkpoint` in the internal implementation of `BaseCheckpointSaver`.